### DISCUSSION TITLE
Can ast-grep be used as an AST matching tool?

### DISCUSSION THREAD

**Discussion by kjy5** [ast-grep/ast-grep#1340] [Q&A] [answered]
I'm working on a project that uses [GumTreeDiff](https://github.com/GumTreeDiff/gumtree) (Java) to do node mappings or matching between two ASTs. I want to explore a Rust version of my project and ast-grep appears to be the library that would do what GumTree does for me. However, I couldn't find specific language in the documentation that says ast-grep can be used to do matches between two trees.

I understand that I can "search" for a node from one tree to another, but it's unclear if this is the same behavior as doing a 1:1 matching between the nodes of two trees.

Can ast-grep do this kind of 1:1 matching? And if not, what are some Rust libraries that can?

**Comment by HerringtonDarkholme** [comment, marked as answer]
I believe difftastic https://github.com/Wilfred/difftastic should be a better fit for AST matching/comparing.

**Reply by kjy5** [reply]
Thanks! I'll look into them. It seems to be a CLI tool though as opposed to a library. Do you know if it can be used as a library in code?

**Reply by HerringtonDarkholme** [reply]
You can take a look at https://github.com/Wilfred/difftastic/issues/216

**Reply by kjy5** [reply]
Hmm looks like it doesn't exist yet. Just to check, ast-grep not able to do this 1:1 mapping or are you suggesting there are probably better options I should use instead?

**Reply by HerringtonDarkholme** [reply]
I'm not 100% sure about what "AST matching" you are referring to. If you are talking about getting the diff of two AST, that will be out of ast-grep's scope. See https://difftastic.wilfred.me.uk/tree_diffing.html and also Wilfred's blogpost.

If you are finding a crate to match if two trees are exactly the same, then you can trivially write a tree-traversal by using ast-grep's `Node` struct.

**Reply by kjy5** [reply]
By AST matching I'm mean given two ASTs, tell me which node from one corresponds to another. There should be a 1:1 matching between these two trees and any extras will be considered unmatched.

**Reply by HerringtonDarkholme** [reply]
> one corresponds to another. There should be a 1:1 matching between these two trees and any extras will be considered 

I would say ast-grep can definitely do that. You can also use tree-sitter directly and it shouldn't be too hard to write a comparison function from scratch.

**Reply by kjy5** [reply]
Nice! Thank you.
